File: examples/spacetrack/constellation_query.py
# /// script
# dependencies = ["brahe", "numpy"]
# ///
"""
Demonstrates querying satellite constellations from SpaceTrack.
Uses the ICEYE constellation as a smaller example.
"""

import os

import numpy as np

import brahe as bh

# Initialize EOP data for propagation
bh.initialize_eop()

# Get credentials from environment
# Note: For production use, most users should use the default constructor:
#   client = bh.SpaceTrackClient(user, password)
# These examples use the test server for automated testing.
user = os.environ.get("TEST_SPACETRACK_USER", "")
password = os.environ.get("TEST_SPACETRACK_PASS", "")

if not user or not password:
    print("Set TEST_SPACETRACK_USER and TEST_SPACETRACK_PASS environment variables")
else:
    client = bh.SpaceTrackClient(
        user, password, base_url="https://for-testing-only.space-track.org/"
    )

    # Query ICEYE constellation (smaller than Starlink, good for demo)
    # First query GP records with pattern matching
    iceye_records = client.gp(
        object_name="~~ICEYE%",
        limit=10,
    )

    # Convert GP records to propagators using embedded TLE data
    iceye_props = []
    for r in iceye_records:
        prop = bh.SGPPropagator.from_3le(
            r.object_name, r.tle_line1, r.tle_line2, step_size=60.0
        )
        iceye_props.append(prop)

    print(f"Retrieved {len(iceye_props)} ICEYE propagators")

    # Calculate orbital statistics
    epoch = bh.Epoch.from_datetime(2024, 6, 1, 12, 0, 0.0, 0.0, bh.TimeSystem.UTC)
    altitudes = []

    for prop in iceye_props:
        prop.propagate_to(epoch)
        state = prop.current_state()
        r = np.sqrt(state[0] ** 2 + state[1] ** 2 + state[2] ** 2)
        altitudes.append((r - bh.R_EARTH) / 1000)

    print(f"Altitude range: {min(altitudes):.1f} - {max(altitudes):.1f} km")
    print(f"Mean altitude: {np.mean(altitudes):.1f} km")
    # Retrieved 10 ICEYE propagators
    # Altitude range: 560.2 - 580.8 km
    # Mean altitude: 570.3 km

print("Example completed successfully!")

File: examples/spacetrack/constellation_query.rs
//! Demonstrates querying satellite constellations from SpaceTrack.
//! Uses the ICEYE constellation as a smaller example.
//!
//! FLAGS = ["CI-ONLY"]

use brahe as bh;
use brahe::propagators::SGPPropagator;
use brahe::spacetrack::{BlockingSpaceTrackClient, TEST_BASE_URL};
use brahe::traits::SStatePropagator;

fn main() {
    bh::initialize_eop().unwrap();

    // Get credentials from environment
    // Note: For production use, most users should use the default constructor:
    //   BlockingSpaceTrackClient::new("user", "password")
    // These examples use the test server for automated testing.
    let user = std::env::var("TEST_SPACETRACK_USER").unwrap_or_default();
    let password = std::env::var("TEST_SPACETRACK_PASS").unwrap_or_default();

    if user.is_empty() || password.is_empty() {
        println!("Set TEST_SPACETRACK_USER and TEST_SPACETRACK_PASS environment variables");
    } else {
        let client = BlockingSpaceTrackClient::with_base_url(&user, &password, TEST_BASE_URL)
            .expect("Failed to create client");

        // Query ICEYE constellation (smaller than Starlink, good for demo)
        // First query GP records with pattern matching
        let iceye_records = client
            .gp(
                None,
                Some("~~ICEYE%"),
                None,
                None,
                None,
                None,
                Some(10),
                None,
            )
            .expect("Query failed");

        // Convert GP records to propagators using embedded TLE data
        let mut iceye_props: Vec<SGPPropagator> = Vec::new();
        for r in &iceye_records {
            if let (Some(name), Some(line1), Some(line2)) = (
                r.object_name.as_ref(),
                r.tle_line1.as_ref(),
                r.tle_line2.as_ref(),
            ) {
                if let Ok(prop) = SGPPropagator::from_3le(Some(name), line1, line2, 60.0) {
                    iceye_props.push(prop);
                }
            }
        }

        println!("Retrieved {} ICEYE propagators", iceye_props.len());

        // Calculate orbital statistics
        let epoch = bh::Epoch::from_datetime(2024, 6, 1, 12, 0, 0.0, 0.0, bh::TimeSystem::UTC);
        let mut altitudes: Vec<f64> = Vec::new();

        for prop in &mut iceye_props {
            prop.propagate_to(epoch);
            let state = prop.current_state();
            let r = (state[0].powi(2) + state[1].powi(2) + state[2].powi(2)).sqrt();
            altitudes.push((r - bh::R_EARTH) / 1000.0);
        }

        if !altitudes.is_empty() {
            let min_alt = altitudes.iter().cloned().fold(f64::INFINITY, f64::min);
            let max_alt = altitudes.iter().cloned().fold(f64::NEG_INFINITY, f64::max);
            let mean_alt: f64 = altitudes.iter().sum::<f64>() / altitudes.len() as f64;

            println!("Altitude range: {:.1} - {:.1} km", min_alt, max_alt);
            println!("Mean altitude: {:.1} km", mean_alt);
        }
        // Retrieved 10 ICEYE propagators
        // Altitude range: 560.2 - 580.8 km
        // Mean altitude: 570.3 km
    }

    println!("Example completed successfully!");
}
